ml-algorithms/src/main/java/org/opensearch/ml/engine/encryptor/EncryptorImpl.java

```diff
@@ -104,28 +113,67 @@ private void initMasterKey() {
+                                try (ThreadContext.StoredContext threadContext = client.threadPool().getThreadContext().stashContext()) {
+                                    client.get(getMasterKeyRequest, ActionListener.wrap(getMasterKeyResponse -> {
+                                        if (getMasterKeyResponse != null && getMasterKeyResponse.isExists()) {
+                                            final String masterKey = (String) getMasterKeyResponse.getSourceAsMap().get(MASTER_KEY);
+                                            this.masterKey = masterKey;
+                                            log.info("ML encryption master key already initialized, no action needed");
+                                            latch.countDown();
+                                        } else {
+                                            exceptionRef.set(new ResourceNotFoundException(MASTER_KEY_NOT_READY_ERROR));
+                                            latch.countDown();
+                                        }
+                                    }, error -> {
+                                        log.debug("Failed to get ML encryption master key", e);
+                                        exceptionRef.set(error);
+                                        latch.countDown();
+                                    }));
+                                }
+                            } else {
+                                log.debug("Failed to index ML encryption master key", e);
+                                exceptionRef.set(e);
+                                latch.countDown();
+                            }
+                        }));
                     } else {
-                        exceptionRef.set(new ResourceNotFoundException(MASTER_KEY_NOT_READY_ERROR));
+                        final String masterKey = (String) getResponse.getSourceAsMap().get(MASTER_KEY);
+                        this.masterKey = masterKey;
+                        log.info("ML encryption master key already initialized, no action needed");
+                        latch.countDown();
                     }
                 }, e -> {
-                    log.error("Failed to get ML encryption master key", e);
+                    log.debug("Failed to get ML encryption master key from config index", e);
                     exceptionRef.set(e);
-                }), latch), () -> context.restore()));
+                    latch.countDown();
+                }));
             }
-        } else {
-            exceptionRef.set(new ResourceNotFoundException(MASTER_KEY_NOT_READY_ERROR));
+        }, e -> {
+            log.debug("Failed to init ML config index", e);
+            exceptionRef.set(e);
             latch.countDown();
-        }
+        }));
 
         try {
-            latch.await(5, SECONDS);
+            latch.await(1, SECONDS);
```

Comment:
Can we make sure the blocking here is safe? It won't have impact on other threads like event loop?

Reply:
This is not a perfect fix. Just for quickly unblock query assistant. Did testing on single node and multi-node cluster, looks good.

Reply:
@rbhavna  will have a long-term fix which will remove such await

Reply:
@rbhavna please create an issue to track this.